### PR TITLE
Added login logic and local storage

### PR DESCRIPTION
**Pull request by JorgeGarciaIrazabal**
- storing facebook data in python server
- Using local storage to save logged user (This can be use to redirect to other page if it is already logged)
- Created model super class so we have `createAt`, `updatedAt`, and `constructFromJson` in all our models
- Created Singleton class. We should use singleton for our services (easier to tests)

### CODE REVIEW

File: backend/db_context/tables.py [hunk 1]
```diff
@@ -17,20 +17,25 @@ class ReactChatDocument(Document):
...
+        values = {underscore(k): v for k, v in values.items()}
         super().__init__(*args, **values)
 
     def save(self, *args, **kwargs):
-        self.updated_on = datetime.utcnow()
+        self.updated_at = datetime.utcnow()
         return super().save(*args, **kwargs)
 
+    def update(self, **kwargs):
+        kwargs = {underscore(k): v for k, v in kwargs.items()}
```
Comment (JorgeGarciaIrazabal) [inline backend/db_context/tables.py:28]: change camelcase to underscore so we can create model directly from the client data

File: backend/db_context/tables.py [hunk 2]
```diff
@@ -42,8 +47,4 @@ class Message(ReactChatDocument):
 
 if __name__ == '__main__':
     connect('ReactChat')
```
Comment (JorgeGarciaIrazabal) [inline backend/db_context/tables.py:49]: This is not executed normally, but you can run now this file to remove every user form the DataBase (usefull for development)

File: backend/hubs/UserHub.py [hunk 3]
```diff
@@ -7,15 +7,15 @@
...
-            user = User.objects(email=user.email).get()
+            user = User.objects(email=user_data['email']).get()
 
         utils_hub = UtilsAPIHub.get_instance()
         ''' :type :UtilsAPIHub'''
-        utils_hub.set_id(user.pk, _sender)
+        utils_hub.set_id(str(user.pk), _sender)
 
-        return user.pk
+        return user
```
Comment (JorgeGarciaIrazabal) [inline backend/hubs/UserHub.py:21]: returning the whole user so we can get `createdAt` and `updatedAt` also

File: backend/server.py [hunk 4]
```diff
@@ -10,6 +11,14 @@
 
 from db_context.tables import ReactChatDocument
 
+
+def get_ip():
```
Comment (JorgeGarciaIrazabal) [inline backend/server.py:15]: we get the ip just to be shown in console

File: src/components/FacebookLoginButton.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
+import React, {Component, PropTypes} from 'react';
```
Comment (JorgeGarciaIrazabal) [inline src/components/FacebookLoginButton.js:1]: created a `FacebookLoginButton` component to handle all our specific logic (what permissions, what parameter to get, etc)

File: src/hubsApi.js [hunk 6]
```diff
@@ -233,122 +233,122 @@ function HubsAPI(serverTimeout, wsClientClass, PromiseClass) {
         return promise;
```
Comment (JorgeGarciaIrazabal) [inline src/hubsApi.js:233]: Auto-generated file, don't worry about this :stuck_out_tongue_closed_eyes:

File: src/models/contact.js [hunk 7]
```diff
@@ -1,23 +0,0 @@
-import {observable} from 'mobx';
```
Comment (JorgeGarciaIrazabal) [inline src/models/contact.js:1]: Contact is not needed (I don't know why I created it)

File: src/models/message.js [hunk 8]
```diff
@@ -1,7 +1,9 @@
 import {observable} from 'mobx';
```
Comment (JorgeGarciaIrazabal) [inline src/models/message.js:1]: Basically changes to make Message inherit from Model

File: src/models/model.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import {observable} from 'mobx';
```
Comment (JorgeGarciaIrazabal) [inline src/models/model.js:1]: new base class Model

File: src/models/user.js [hunk 10]
```diff
@@ -1,24 +1,23 @@
 import {observable} from 'mobx';
```
Comment (JorgeGarciaIrazabal) [inline src/models/user.js:1]: same as Message

File: src/models/model.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    Object.assign(model, camelCasedJson);
+    return model;
+  }
+
+  toJson(){
+    return JSON.stringify(this);
+  }
+
+  static _toCamelCase(underscoreName: string){
```
Comment (JorgeGarciaIrazabal) [inline src/models/model.js:31]: changing underscored var names for camelCase so we can use directly the data form the server

File: src/pages/login.js [hunk 12]
```diff
@@ -45,37 +59,10 @@ class Login extends Component {
...
-                    null,
-                    this._responseInfoCallback,
-                  );
-                  new GraphRequestManager().addRequest(infoRequest).start();
-                }
-              }
-            }
-            onLogoutFinished={() => alert("logout.")}/>
-        </View>
+        <FacebookLoginButton onUserFetched={this.logIn.bind(this)}/>
```
Comment (JorgeGarciaIrazabal) [inline src/pages/login.js:62]: use the new `FacebookLoginButton` component instead of creating it here

File: src/services/api.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+import HubsApi from '../hubsApi';
```
Comment (JorgeGarciaIrazabal) [inline src/services/api.js:1]: API service to communicate with the server

File: src/services/localStorage.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
+import {AsyncStorage} from 'react-native'
```
Comment (JorgeGarciaIrazabal) [inline src/services/localStorage.js:1]: LocalStorage class to get and set persistent data

File: src/services/singleton.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
+const singletonEnforcer = Symbol();
```
Comment (JorgeGarciaIrazabal) [inline src/services/singleton.js:1]: simple class to fore services to be singleton

File: src/services/api.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+export default class API {
+  /** @type HubsAPI */
+  static hubsApi = null;
+
+  static async setup() {
+    API.hubsApi = new HubsApi(10000);
+    await API.hubsApi.connect('ws://192.168.0.15:8844/');
```
Comment (JorgeGarciaIrazabal) [inline src/services/api.js:9]: change the `ip` for the one shown in the console where you are running the server (in the future we will need to create a script to automatically update this value)

**Review by Fdom92 — APPROVED** [review]
Nice changes 👍

**Pull request merged**